### PR TITLE
[SPARK-27976][SQL] Add built-in Array Functions: array_append

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Function | Return Type | Description | Example | Result
-- | -- | -- | -- | --
array_append(anyarray,anyelement) | anyarray | append an element to the end of an array | array_append(ARRAY[1,2], 3) | {{ {1,2,3}}}

### Why are the changes needed?

https://www.postgresql.org/docs/current/functions-array.html

Other DBs:
https://phoenix.apache.org/language/functions.html#array_append
https://docs.teradata.com/reader/kmuOwjp1zEYg98JsB8fu_A/68fdFR3LWhx7KtHc9Iv5Qg

### Does this PR introduce any user-facing change?

yes. add a new array function array_append

### How was this patch tested?

add uts

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu @wangyum

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 1]
```diff
@@ -3928,3 +3928,88 @@ case class ArrayExcept(left: Expression, right: Expression) extends ArrayBinaryL
...
+  _FUNC_(array, element) - Returns an array of appending an element to the end of an array
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(1, 2, 3), 3);
+       [1,2,3,3]
+  """,
+  since = "3.0.0")
+case class ArrayAppend(left: Expression, right: Expression) extends BinaryExpression {
+
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: You need to handle `nullable` and `foldable`, too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 2]
```diff
@@ -3928,3 +3928,88 @@ case class ArrayExcept(left: Expression, right: Expression) extends ArrayBinaryL
...
+         |for (int $i = 0; i < $inputArray.numElements(); $i ++) {
+         |  $assignment
+         |}
+         |$setNewValue
+         |${ev.value} = $newArray;
+         |""".stripMargin
+    })
+  }
+
+  override def dataType: DataType = left.dataType match {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: You need to take care of `right.nullable`.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: all checked in `checkInputDataTypes`

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: I got your point here. thanks.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Yea, thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 3]
```diff
@@ -3928,3 +3928,128 @@ case class ArrayExcept(left: Expression, right: Expression) extends ArrayBinaryL
...
+       [1,2,3,3]
+      > SELECT _FUNC_(array(1, 2, 3), null);
+       [1,2,3,null]
+      > SELECT _FUNC_(a, e) FROM VALUES (array(1,2), 3), (array(3, 4), null), (null, 3) tbl(a, e);
+       NULL
+       [1,2,3]
+       [3,4,null]
+  """,
+  since = "3.0.0")
+case class ArrayAppend(left: Expression, right: Expression) extends BinaryExpression {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: This is equivalent to Concat(Seq(left + CreateArray(Seq(right)))). Shall we use existing expressions instead adding duplicate codes?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 1 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: @yaooqinn You forgot to override `foldable`...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: There are some cases result different.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Besides the Concat function for array is completely different from Postgres, I think that we have it implemented wrong

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Could you please put the case here?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]:
```sql
postgres=# select array_cat(array[1,2], array[2]);
 array_cat
-----------
 {1,2,2}
(1 row)

postgres=# select concat(array[1,2], array[2]);
  concat
----------
 {1,2}{2}
(1 row)
```

```sql
select concat(array(1,2), array(2));
[1,2,2]
```

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Some bugs, also
```sql
spark-sql> select concat(null, array(1));
Error in query: cannot resolve 'concat(CAST(NULL AS STRING), array(1))' due to data type mismatch: input to function concat should all be the same type, but it's [string, array<int>]; line 1 pos 7;
'Project [unresolvedalias(concat(cast(null as string), array(1)), None)]
+- OneRowRelation

spark-sql>
```

```
spark-sql> select concat(array(1), array(null));
19/11/04 10:26:33 ERROR SparkSQLDriver: Failed in [select concat(array(1), array(null))]
scala.MatchError: NullType (of class org.apache.spark.sql.types.NullType$)
	at org.apache.spark.sql.catalyst.expressions.Cast.castToInt(Cast.scala:434)
	at org.apache.spark.sql.catalyst.expressions.Cast.org$apache$spark$sql$catalyst$expressions$Cast$$cast(Cast.scala:622)
	at org.apache.spark.sql.catalyst.expressions.Cast.castArray(Cast.scala:563)
	at org.apache.spark.sql.catalyst.expressions.Cast.org$apache$spark$sql$catalyst$expressions$Cast$$cast(Cast.scala:627)
	at org.apache.spark.sql.catalyst.expressions.Cast.cast$lzycompute(Cast.scala:639)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: the foldable here is the same as `BinaryExpression`, what about we just inherit it？

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 3 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]:
```
spark-sql> select concat(null, array(1));
```
Is this valid in `ArrayAppend`? I think `checkInputDataTypes` just throws an exception in the case.

```
spark-sql> select concat(array(1), array(null));
```
What's the spark version that you used? I tried in the current master;
```
scala> sql("""select concat(array(1), array(null))""").show()
+-----------------------------+
|concat(array(1), array(NULL))|
+-----------------------------+
|                         [1,]|
+-----------------------------+
```

Anyway, I think you just need to handle these error cases for ArrayAppend.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: > ```sql
> postgres=# select array_cat(array[1,2], array[2]);
>  array_cat
> -----------
>  {1,2,2}
> (1 row)
> 
> postgres=# select concat(array[1,2], array[2]);
>   concat
> ----------
>  {1,2}{2}
> (1 row)
> ```
> 
> ```sql
> select concat(array(1,2), array(2));
> [1,2,2]
> ```

This looks like Spark concat for array type is like Presto's array_cat.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]:
```
-- !query 13
select array_append(array(1,2), null)
-- !query 13 schema
struct<array_append(array(1, 2), NULL):array<int>>
-- !query 13 output
[1,2,null]
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 1 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Ur, I see. I missed this is a binary expr. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Sorry I made a mistake of the cases, i will recheck the master branch

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: I just have a quick research on the master branch which I build on 2019-10-29, 

spark-sql> select concat(array(null), null);
Error in query: cannot resolve 'concat(array(NULL), CAST(NULL AS STRING))' due to data type mismatch: input to function concat should all be the same type, but it's [array<null>, string]; line 1 pos 7;
'Project [unresolvedalias(concat(array(null), cast(null as string)), None)]
+- OneRowRelation

spark-sql> select concat(array(1, 2), null);
Error in query: cannot resolve 'concat(array(1, 2), CAST(NULL AS STRING))' due to data type mismatch: input to function concat should all be the same type, but it's [array<int>, string]; line 1 pos 7;
'Project [unresolvedalias(concat(array(1, 2), cast(null as string)), None)]
+- OneRowRelation

spark-sql> select concat(array(1, 2), array(null));
[1,2,null]
Time taken: 2.908 seconds, Fetched 1 row(s)
```

Also, currently I don't know how to change `select array_append(null, 3);` this to concat

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: > > ```sql
> > postgres=# select array_cat(array[1,2], array[2]);
> >  array_cat
> > -----------
> >  {1,2,2}
> > (1 row)
> > 
> > postgres=# select concat(array[1,2], array[2]);
> >   concat
> > ----------
> >  {1,2}{2}
> > (1 row)
> > ```
> > 
> > 
> > ```sql
> > select concat(array(1,2), array(2));
> > [1,2,2]
> > ```
> 
> This looks like Spark concat for array type is like Presto's array_cat. May not means it is wrong.

Concat used to be a string function as Hive before 2.4, and I do not know which standard or practice it follows now

**Comment by yaooqinn** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Hi @maropu , the key difference is when the array is null and the new element is not, array_append will create new array but concat's result will be null. please see the following cases.
```
-- !query 18
select array_append(a, e) from VALUES (array(1,2), 3), (array(3, 4), null), (null, 3), (null, null) tbl(a, e)
-- !query 18 schema
struct<array_append(a, e):array<int>>
-- !query 18 output
NULL
[1,2,3]
[3,4,null]
[3]

-- !query 19
select concat(a, array(e)) from VALUES (array(1,2), 3), (array(3, 4), null), (null, 3), (null, null) tbl(a, e)
-- !query 19 schema
struct<concat(a, array(e)):array<int>>
-- !query 19 output
NULL
NULL
[1,2,3]
[3,4,null]
```

**Comment by yaooqinn** [comment]
```java
--- !query 24
-explain codegen select array_append(a, e) from VALUES (array(1,2), 3), (array(3, 4), null), (null, 3), (null, null) tbl(a, e)
--- !query 24 schema
-struct<plan:string>
--- !query 24 output
-Found 1 WholeStageCodegen subtrees.
-== Subtree 1 / 1 (maxMethodCodeSize:453; maxConstantPoolSize:151(0.23% used); numInnerClasses:0) ==
-*Project [array_append(a#x, e#x) AS array_append(a, e)#x]
-+- *LocalTableScan [a#x, e#x]
-
-Generated code:
-/* 001 */ public Object generate(Object[] references) {
-/* 002 */   return new GeneratedIteratorForCodegenStage1(references);
-/* 003 */ }
-/* 004 */
-/* 005 */ // codegenStageId=1
-/* 006 */ final class GeneratedIteratorForCodegenStage1 extends org.apache.spark.sql.execution.BufferedRowIterator {
-/* 007 */   private Object[] references;
-/* 008 */   private scala.collection.Iterator[] inputs;
-/* 009 */   private scala.collection.Iterator localtablescan_input_0;
-/* 010 */   private org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter[] project_mutableStateArray_0 = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter[1];ssions.codegen.UnsafeArrayWriter[] project_mutableStateArray_1 = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeArrayWriter[1];
-/* 013 */   public GeneratedIteratorForCodegenStage1(Object[] references) {
-/* 014 */     this.references = references;
-/* 015 */   }
-/* 016 */
-/* 017 */   public void init(int index, scala.collection.Iterator[] inputs) {
-/* 018 */     partitionIndex = index;
-/* 019 */     this.inputs = inputs;
-/* 020 */     localtablescan_input_0 = inputs[0];
-/* 021 */     project_mutableStateArray_0[0] = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter(1, 32);
-/* 022 */     project_mutableStateArray_1[0] = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeArrayWriter(project_mutableStateArray_0[0], 4);
-/* 024 */   }
-/* 025 */
-/* 026 */   protected void processNext() throws java.io.IOException {
-/* 027 */     while ( localtablescan_input_0.hasNext()) {
-/* 028 */       InternalRow localtablescan_row_0 = (InternalRow) localtablescan_input_0.next();
-/* 029 */       ((org.apache.spark.sql.execution.metric.SQLMetric) references[0] /* numOutputRows */).add(1);
-/* 030 */       boolean project_isNull_0 = true;
-/* 031 */       ArrayData project_value_0 = null;
-/* 032 */       boolean localtablescan_isNull_0 = localtablescan_row_0.isNullAt(0);
-/* 033 */       ArrayData localtablescan_value_0 = localtablescan_isNull_0 ?
-/* 034 */       null : (localtablescan_row_0.getArray(0));
-/* 035 */       boolean localtablescan_isNull_1 = localtablescan_row_0.isNullAt(1);
-/* 036 */       int localtablescan_value_1 = localtablescan_isNull_1 ?
-/* 037 */       -1 : (localtablescan_row_0.getInt(1));
-/* 038 */       if (localtablescan_isNull_0) {
-/* 039 */         if (!localtablescan_isNull_1) {
-/* 040 */           project_isNull_0 = false; // resultCode could change nullability.
-/* 041 */           int project_oldArraySize_0 = 0;
-/* 042 */
-/* 043 */           ArrayData project_newArray_0 = ArrayData.allocateArrayData(
-/* 044 */             4, 1, "array_append failed");
-/* 045 */
-/* 046 */           if (localtablescan_isNull_1) {
-/* 047 */             project_newArray_0.setNullAt(project_oldArraySize_0);
-/* 048 */           } else {
-/* 049 */             project_newArray_0.setInt(project_oldArraySize_0, localtablescan_value_1);
-/* 050 */           }
-/* 051 */
-/* 052 */           project_value_0 = project_newArray_0;
-/* 053 */         }
-/* 054 */       } else {
-/* 055 */         project_isNull_0 = false; // resultCode could change nullability.
-/* 056 */
-/* 057 */         int project_oldArraySize_0 = localtablescan_value_0.numElements();
-/* 058 */         int project_newArraySize_0 = project_oldArraySize_0 + 1;
-/* 059 */
-/* 060 */         ArrayData project_newArray_0 = ArrayData.allocateArrayData(
-/* 061 */           4, project_newArraySize_0, "array_append failed");
-/* 062 */
-/* 063 */         for (int project_i_0 = 0; project_i_0 < localtablescan_value_0.numElements(); project_i_0 ++) {
-/* 064 */           if (localtablescan_value_0.isNullAt(project_i_0)) {
-/* 065 */             project_newArray_0.setNullAt(project_i_0);
-/* 066 */           } else {
-/* 067 */             project_newArray_0.setInt(project_i_0, localtablescan_value_0.getInt(project_i_0));
-/* 068 */           }
-/* 069 */
-/* 070 */         }
-/* 071 */
-/* 072 */         if (localtablescan_isNull_1) {
-/* 073 */           project_newArray_0.setNullAt(project_oldArraySize_0);
-/* 074 */         } else {
-/* 075 */           project_newArray_0.setInt(project_oldArraySize_0, localtablescan_value_1);
-/* 076 */         }
-/* 077 */
-/* 078 */         project_value_0 = project_newArray_0;
-/* 079 */
-/* 080 */       }
-/* 081 */       project_mutableStateArray_0[0].reset();
-/* 082 */
-/* 083 */       project_mutableStateArray_0[0].zeroOutNullBytes();
-/* 084 */
-/* 085 */       if (project_isNull_0) {
-/* 086 */         project_mutableStateArray_0[0].setNullAt(0);
-/* 087 */       } else {
-/* 088 */         // Remember the current cursor so that we can calculate how many bytes are
-/* 089 */         // written later.
-/* 090 */         final int project_previousCursor_0 = project_mutableStateArray_0[0].cursor();
-/* 091 */
-/* 092 */         final ArrayData project_tmpInput_0 = project_value_0;
-/* 093 */         if (project_tmpInput_0 instanceof UnsafeArrayData) {
-/* 094 */           project_mutableStateArray_0[0].write((UnsafeArrayData) project_tmpInput_0);
-/* 095 */         } else {
-/* 096 */           final int project_numElements_0 = project_tmpInput_0.numElements();
-/* 097 */           project_mutableStateArray_1[0].initialize(project_numElements_0);
-/* 098 */
-/* 099 */           for (int project_index_0 = 0; project_index_0 < project_numElements_0; project_index_0++) {
-/* 100 */             if (project_tmpInput_0.isNullAt(project_index_0)) {
-/* 101 */               project_mutableStateArray_1[0].setNull4Bytes(project_index_0);
-/* 102 */             } else {
-/* 103 */               project_mutableStateArray_1[0].write(project_index_0, project_tmpInput_0.getInt(project_index_0));
-/* 104 */             }
-/* 105 */
-/* 106 */           }
-/* 107 */         }
-/* 108 */
-/* 109 */         project_mutableStateArray_0[0].setOffsetAndSizeFromPreviousCursor(0, project_previousCursor_0);
-/* 110 */       }
-/* 111 */       append((project_mutableStateArray_0[0].getRow()));
-/* 112 */       if (shouldStop()) return;
-/* 113 */     }
-/* 114 */   }
-/* 115 */
-/* 116 */ }
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: This is what postgres results：
```SQL
postgres=# select array_append(a, e) from (VALUES (array[1,2], 3), (array[3, 4], null), (null, 3), (null, null)) tbl(a, e);
 array_append
--------------
 {1,2,3}
 {3,4,NULL}
 {3}
 {NULL}
(4 rows)
```

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: But, I personally think it is more natural that the result is null if the left input (array) is null. WDYT? @viirya @cloud-fan

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: yea, I think null is not an empty array, so appending something to a null (left) sounds to be a null too.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: also cc @HyukjinKwon @gatorsmile @wangyum @dongjoon-hyun

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: I don't have a strong preference. Seems fine to follow postgres here, i.e. `array_append` is not nullable.

**Comment by yaooqinn** [comment]
@cloud-fan @maropu @viirya @dongjoon-hyun Can we reach an agreement here？ Thanks.

**Comment by maropu** [comment]
I just want to close this for now and resolve the jira as "later" because we have many open PRs with high priorities for the v3.0 release. WDYT, @yaooqinn ? Anyway, thanks for the work!

**Comment by yaooqinn** [comment]
OK, thanks @maropu.

**Pull request closed without merging**